### PR TITLE
Optimize the number of lines in the output when a type matching error…

### PR DESCRIPTION
**Pull request by qianmoQ**
… occurs

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -3125,7 +3126,12 @@ private void coerceType(Expression expression, Type actualType, Type expectedTyp
         {
             if (!actualType.equals(expectedType)) {
                 if (!typeCoercion.canCoerce(actualType, expectedType)) {
-                    throw semanticException(TYPE_MISMATCH, expression, "%s must evaluate to a %s (actual: %s)", message, expectedType, actualType);
+                    if (expression.getLocation().isPresent()) {
+                        NodeLocation location =  expression.getLocation().get();
+                        throw semanticException(TYPE_MISMATCH, expression, "%s must evaluate to a %s (actual: %s) line: %s column: %s", message, expectedType, actualType, location.getLineNumber(), location.getColumnNumber());
+                    } else {
```
[the remark this answers is not in the document]
Comment (qianmoQ) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java]: Fixed

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -3125,7 +3126,12 @@ private void coerceType(Expression expression, Type actualType, Type expectedTyp
         {
             if (!actualType.equals(expectedType)) {
                 if (!typeCoercion.canCoerce(actualType, expectedType)) {
-                    throw semanticException(TYPE_MISMATCH, expression, "%s must evaluate to a %s (actual: %s)", message, expectedType, actualType);
+                    if (expression.getLocation().isPresent()) {
+                        NodeLocation location =  expression.getLocation().get();
```
[the remark this answers is not in the document]
Comment (qianmoQ) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java]: Fixed

**Comment by krvikash** [comment]
Hi @qianmoQ, I could see that the error already contains (`line 1:149:`) the line number and column number at the beginning of the error message. Why do you need it again in the message?

```
line 1:149: Function JSON_VALUE default ON EMPTY result must evaluate to a double (actual: varchar(4))
```

https://github.com/trinodb/trino/blob/master/core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java#L5913

**Comment by qianmoQ** [comment]
> 

The error message here was not submitted by me. I think it was submitted here in 2021

**Comment by qianmoQ** [comment]
> Without your change The error message is
> 
> ```
> line 1:149: Function JSON_VALUE default ON EMPTY result must evaluate to a double (actual: varchar(4))
> ```
> 
> With your change The error message is
> 
> ```
> line 1:149: Function JSON_VALUE default ON EMPTY result must evaluate to a double (actual: varchar(4)) line: 1 column: 149"
> ```
> 
> I think `line: 1 column: 149` is redundant here. we already have `line 1:149` with the error message. My question here is why do you need `line: %d column: %d` when we have already `line 1:149`?

Maybe it is the problem of my version. The error message I saw did not have the feedback line number, so I added it

**Comment by qianmoQ** [comment]
> Without your change The error message is
> 
> ```
> line 1:149: Function JSON_VALUE default ON EMPTY result must evaluate to a double (actual: varchar(4))
> ```
> 
> With your change The error message is
> 
> ```
> line 1:149: Function JSON_VALUE default ON EMPTY result must evaluate to a double (actual: varchar(4)) line: 1 column: 149"
> ```
> 
> I think `line: 1 column: 149` is redundant here. we already have `line 1:149` with the error message. My question here is why do you need `line: %d column: %d` when we have already `line 1:149`?

The error I got was:

```
Query 20230220_070617_00011_skthd failed:  Function $internal$dynamic_filter_function(bigint, varchar, varchar,  boolean):boolean argument 0 must evaluate to a bigint (actual: varchar)
```

**Comment by krvikash** [comment]
What trino version are you using?

**Comment by qianmoQ** [comment]
> What trino version are you using?

image

**Comment by krvikash** [comment]
Tests are failing, Please take a look. Squash commits in one single commit.

**Pull request closed without merging**